sdk/python/lib/pulumi/output.py

```diff
@@ -253,27 +253,32 @@ def from_input(val: Input[T]) -> 'Output[T]':
         if isinstance(val, dict):
             # Since Output.all works on lists early, serialize this dictionary into a list of lists first.
             # Once we have a output of the list of properties, we can use an apply to re-hydrate it back into a dict.
-            transformed_items = [[k, Output.from_input(v)] for k, v in val.items()]
-            return Output.all(*transformed_items).apply(lambda props: {k: v for k, v in props}, True)
+            dict_items = [[k, Output.from_input(v)] for k, v in val.items()]
+            # type checker doesn't like returing a Dict in the apply callback
+            fn = cast(Callable[[List[Any]], T], lambda props: {k: v for k, v in props}) # pylint: disable=unnecessary-comprehension
+            return Output.all(*dict_items).apply(fn, True)
 
         if isinstance(val, list):
-            transformed_items = [Output.from_input(v) for v in val]
-            return Output.all(*transformed_items)
+            list_items: List[Union[Any, Awaitable[Any], Output[Any]]] = [Output.from_input(v) for v in val]
+            # invariant: http://mypy.readthedocs.io/en/latest/common_issues.html#variance
+            output: Output[T] = cast(Output[T], Output.all(*list(list_items))) # type: ignore
+            return output
 
         # If it's not an output, list, or dict, it must be known and not secret
-        is_known_fut = asyncio.Future()
-        is_secret_fut = asyncio.Future()
+        is_known_fut: asyncio.Future[bool] = asyncio.Future()
+        is_secret_fut: asyncio.Future[bool] = asyncio.Future()
```

Comment:
so how does this line work if python is invariant as you mentioned?  what is the return type of `asyncio.Future()`?

Reply:
I don't know exactly the rules for how this works - but I think the type actually gets inferred from the context here (like TS contextual typing).  

Future is indeed invariant - see:

https://github.com/python/typeshed/blob/c32e1e228038af9fe95e5899431cd442a5772ad6/stdlib/3/concurrent/futures/_base.pyi#L26